packages/styled/types/tests.tsx

```diff
@@ -217,3 +217,69 @@ const Input5 = styled.input`
   // $ExpectError
   ;<StyledCompWithoutAs as={Section} />
 }
+
+{
+  // Props forwarding through StyledOptions and FilteringStyledOptions
+
+  const fc: React.FC<{ foo: string }> = props => <div {...props} />
+
+  styled(fc, { shouldForwardProp: (prop: 'foo') => true })({})
+
+  styled(fc, { shouldForwardProp: (prop: string) => true })({})
+
+  // $ExpectError
+  styled(fc, { shouldForwardProp: (prop: 'bar') => true })({})
+
+  const shouldForwardProp1 = (prop: 'foo') => true
+  styled(fc, { shouldForwardProp: shouldForwardProp1 })({})
+
+  const shouldForwardProp2: StyledOptions['shouldForwardProp'] = (
+    prop: 'unknown'
+  ) => true
+  styled(fc, { shouldForwardProp: shouldForwardProp2 })({})
+
+  const shouldForwardProp3 = (prop: 'unknown') => true
+  // $ExpectError
+  styled(fc, { shouldForwardProp: shouldForwardProp3 })({})
+
+  // $ExpectError
+  const shouldForwardProp4: StyledOptions<{
+    foo: string
+  }>['shouldForwardProp'] = (prop: 'unknown') => true
+
+  const shouldForwardProp5 = (prop: 'foo'): prop is 'foo' => true
+  styled(fc, { shouldForwardProp: shouldForwardProp5 })({})
+
+  const shouldForwardProp6: FilteringStyledOptions['shouldForwardProp'] = (
+    prop: 'foo'
+  ): prop is 'foo' => true
+
+  const shouldForwardProp7: FilteringStyledOptions<{
+    foo: string
+  }>['shouldForwardProp'] = (prop: 'foo'): prop is 'foo' => true
+
+  // $ExpectError
+  const shouldForwardProp8: FilteringStyledOptions<{
+    foo: string
+  }>['shouldForwardProp'] = (prop: 'unknown'): prop is 'unknown' => true
+
+  const shouldForwardProp9: FilteringStyledOptions<
+    { foo: string; bar: string },
+    'foo'
+  >['shouldForwardProp'] = (prop: 'foo'): prop is 'foo' => true
+
+  // $ExpectError
+  const shouldForwardProp10: FilteringStyledOptions<
+    { foo: string; bar: string },
+    'foo'
+  >['shouldForwardProp'] = (prop: 'bar'): prop is 'bar' => true
```

Comment:
Should something like this be allowed?
```js
  const shouldForwardProp10: FilteringStyledOptions<
    { foo: string; bar: string },
    'foo'
  >['shouldForwardProp'] = (prop: 'foo' | 'bar'): prop is 'bar' => true
```

Reply:
Seems compatible. The test passes.

Reply:
Hm, even without the proposed patch that would fix the bivariance problem I see this as being an error:
```
Type '(prop: "foo" | "bar") => prop is "bar"' is not assignable to type '(propName: string | number | symbol) => propName is "foo"'.
  Type predicate 'prop is "bar"' is not assignable to 'propName is "foo"'.
    Type '"bar"' is not assignable to type '"foo"'.ts(2322)
```

But perhaps my true question here (I actually don't remember why **exactly** I've asked about this) was about:
```ts
  const shouldForwardProp10: FilteringStyledOptions<
    { foo: string; bar: string },
    'foo'
  >['shouldForwardProp'] = (prop: 'foo' | 'bar'): prop is 'foo' => true
```
and this indeed is OK with the bivariant `shouldForwardProp` but starts erroring if we change this to a regular contravariant function.

Reply:
My fault. I've tested a bit different case.

```
  const shouldForwardProp9: FilteringStyledOptions<
    { foo: string; bar: string },
    'foo'
  >['shouldForwardProp'] = (prop: 'foo' | 'bar'): prop is 'foo' => true

  // $ExpectError
  const shouldForwardProp10: FilteringStyledOptions<
    { foo: string; bar: string },
    'foo'
  >['shouldForwardProp'] = (prop: 'foo' | 'bar'): prop is 'bar' => true
```